templates/base.html

```diff
@@ -115,6 +115,9 @@
     <meta name="msapplication-TileColor" content="{{ BROWSER_COLOR }}">
   {% endif %}
 
+  {% if REL_CANONICAL %}
+    <link rel="canonical" href="{{ SITEURL }}/{{ output_file }}">
```

Comment:
where `output_file` comes from? this looks odd, I would expect something like:

`{{ CANONICAL_URL }}/{{ article.url }}`

What usually changes is the domain or path, not the article URL.

Example:

mysite.com/post/my-cool-article

Canonical could be:

blog.mysite.com/my-cool-article

Reply:
output_file is a Pelican variable as documented [here](https://docs.getpelican.com/en/4.5.0/themes.html#common-variables).  My thinking is that Pelican would in most cases be used to generate the canonical site so my change is simply outputting a canonical link that points to the page being generated with Pelican.  An advantage of doing it this way is that it works for non-articles (e.g. pages).

Reply:
> The name of the file currently being generated. For instance, when Pelican is rendering the home page, output_file will be “index.html”.

This means that home would be mysite.com/index.html, which is wrong, home is mysite.com/, you don't need to append index.html, this is sort of implicit, even the .html, if you have mysite.com/foobar.html just mysite.com/foobar works.

I don't think this is the correct way of implementing canonical urls.

Reply:
I see your point.  I'll try to come up with an improved version in the next few days.  Thanks.

Reply:
I have a couple of different options and would like to know which way you'd like it (and if you think I'm on the right track).  The first option is to place something like the following in base.html:
```
{% if REL_CANONICAL %}
  {% if page %}
    <link rel="canonical" href="{{ SITEURL }}/{{ page.url }}">
  {% elif article %}
    <link rel="canonical" href="{{ SITEURL }}/{{ article.url }}">
  {% elif page_name == 'index' and not articles_previous_page %}
    <link rel="canonical" href="{{ SITEURL }}">
  {% elif author or category or tag or page_name == 'index' %}
    <link rel="canonical" href="{{ SITEURL }}/{{ articles_page.url }}">
  {% else %}
    <link rel="canonical" href="{{ SITEURL }}/{{ output_file }}">
  {% endif %}
{% endif %}
```
The other option is to use overrides in the various templates so base.html would just have:
```
  {% block rel_canonical %}
    {% if REL_CANONICAL %}
      <link rel="canonical" href="{{ SITEURL }}/{{ output_file }}">
    {% endif %}
  {% endblock %}
```
and then the article.html, author.html, category.html, index.html, page.html and tag.html templates would be modified to override the rel_canonical block as appropriate.  For example, the following would be added to article.html:
```
{% block rel_canonical %}
  {% if REL_CANONICAL %}
    <link rel="canonical" href="{{ SITEURL }}/{{ article.url }}">
  {% endif %}
{% endblock %}
```
The same block would be added to page.html except it would use `page.url` instead of `article.url`.  author.html, category.html and tag.html would use `articles_page.url` instead of `article.url`.  To accomplish having SITEURL/index.html have the canonical URL of just SITEURL but still work if the index gets paginated, the following would be added to index.html:
```
{% block rel_canonical %}
  {% if REL_CANONICAL %}
    {% if not articles_previous_page %}
      <link rel="canonical" href="{{ SITEURL }}">
    {% else %}
      <link rel="canonical" href="{{ SITEURL }}/{{ articles_page.url }}">
    {% endif %}
  {% endif %}
{% endblock %}

```
Hopefully, I'm getting closer to meeting expectations but if not, let me know.  Thanks.

Reply:
Good question, probably overriding it looks better, but up to you.